capnslog/pkg_logger.go

```diff
@@ -0,0 +1,237 @@
+
+func (p *packageLogger) Panic(args ...interface{}) {
+	s := fmt.Sprint(args...)
+	p.internalLog(calldepth, CRITICAL, BaseLogEntry(s))
+	panic(s)
+}
+
+func (p *packageLogger) Fatalln(args ...interface{}) {
+	s := fmt.Sprintln(args...)
+	p.internalLog(calldepth, CRITICAL, BaseLogEntry(s))
+	os.Exit(1)
+}
+
+func (p *packageLogger) Fatalf(format string, args ...interface{}) {
+	s := fmt.Sprintf(format, args...)
+	p.internalLog(calldepth, CRITICAL, BaseLogEntry(s))
+	os.Exit(1)
+}
+
+func (p *packageLogger) Fatal(args ...interface{}) {
+	s := fmt.Sprint(args...)
+	p.internalLog(calldepth, CRITICAL, BaseLogEntry(s))
+	os.Exit(1)
+}
+
+// Error Functions
+func (p *packageLogger) Errorln(args ...interface{}) {
+	if p.level < ERROR {
+		return
+	}
+	p.internalLog(calldepth, ERROR, BaseLogEntry(fmt.Sprintln(args...)))
+}
+
+func (p *packageLogger) Errorf(format string, args ...interface{}) {
+	if p.level < ERROR {
+		return
+	}
+	p.internalLog(calldepth, ERROR, BaseLogEntry(fmt.Sprintf(format, args...)))
+}
+
+func (p *packageLogger) Error(entries ...LogEntry) {
+	if p.level < ERROR {
+		return
+	}
+	p.internalLog(calldepth, ERROR, entries...)
+}
+
+func (p *packageLogger) ERROR() bool {
+	return p.level < ERROR
+}
+
+// Warning Functions
+func (p *packageLogger) Warningln(args ...interface{}) {
+	if p.level < WARNING {
+		return
+	}
+	p.internalLog(calldepth, WARNING, BaseLogEntry(fmt.Sprintln(args...)))
+}
+
+func (p *packageLogger) Warningf(format string, args ...interface{}) {
+	if p.level < WARNING {
+		return
+	}
+	p.internalLog(calldepth, WARNING, BaseLogEntry(fmt.Sprintf(format, args...)))
+}
+
+func (p *packageLogger) Warning(entries ...LogEntry) {
+	if p.level < WARNING {
+		return
+	}
+	p.internalLog(calldepth, WARNING, entries...)
+}
+
+func (p *packageLogger) WARNING() bool {
+	return p.level < WARNING
+}
+
+// Notice Functions
+func (p *packageLogger) Noticeln(args ...interface{}) {
+	if p.level < NOTICE {
+		return
+	}
+	p.internalLog(calldepth, NOTICE, BaseLogEntry(fmt.Sprintln(args...)))
+}
+
+func (p *packageLogger) Noticef(format string, args ...interface{}) {
+	if p.level < NOTICE {
+		return
+	}
+	p.internalLog(calldepth, NOTICE, BaseLogEntry(fmt.Sprintf(format, args...)))
+}
+
+func (p *packageLogger) Notice(entries ...LogEntry) {
+	if p.level < NOTICE {
+		return
+	}
+	p.internalLog(calldepth, NOTICE, entries...)
+}
+
+func (p *packageLogger) NOTICE() bool {
+	return p.level < NOTICE
+}
+
+// Info Functions
+func (p *packageLogger) Infoln(args ...interface{}) {
+	if p.level < INFO {
+		return
+	}
+	p.internalLog(calldepth, INFO, BaseLogEntry(fmt.Sprintln(args...)))
+}
+
+func (p *packageLogger) Infof(format string, args ...interface{}) {
+	if p.level < INFO {
+		return
+	}
+	p.internalLog(calldepth, INFO, BaseLogEntry(fmt.Sprintf(format, args...)))
+}
+
+func (p *packageLogger) Info(entries ...LogEntry) {
+	if p.level < INFO {
+		return
+	}
+	p.internalLog(calldepth, INFO, entries...)
+}
+
+func (p *packageLogger) INFO() bool {
```

Comment:
This I've seen: If you're debugging a lot and need to allocate/do-some-work to create a more useful log, you want something like

```go
if log.DEBUG() {
  // fetch some extra statistics
  result := someSlowerOrExpensiveOp()
  log.Debug(result)
}
```

This may only be needed on the lower levels though.